File: snake.py
import numpy as np
import time
import matplotlib.pyplot as plt
from scipy.interpolate import interp2d, CubicSpline


def interp(matrix, xs, ys):
    x = np.arange(matrix.shape[0])
    y = np.arange(matrix.shape[1])
    f = interp2d(x, y, matrix)
    result = f(xs, ys).diagonal()
    return result


def getContours(image: np.ndarray, max_points: int = 50, cmap: str = 'gray') -> np.ndarray:
    '''
    手动标记轮廓
    输入:图像数组,初始点的最大数量,图像的Color Map
    输出:经过插值后的轮廓坐标数组
    '''
    plt.imshow(image, cmap=cmap)
    position = plt.ginput(n=max_points)
    points_num = len(position) + 1
    position.append(position[0])
    position_ori = np.array(position, dtype=float)

    # 对手动标记的点进行3次样条插值
    t_ori = np.arange(1, points_num + 1)
    Spline = CubicSpline(t_ori, position, bc_type='periodic')
    t_new = np.arange(1, points_num+1, 0.1)
    position_new = Spline(t_new)
    position_new = np.array(position_new, dtype=float)

    # 显示标记结果
    plt.imshow(image, cmap='gray')
    plt.plot(position_ori[:, 0], position_ori[:, 1], 'o')
    plt.plot(position_new[:, 0], position_new[:, 1], '-')
    plt.show()

    return position_new


def snake(image: np.ndarray, contour: np.ndarray, alpha: float = 0.2, beta: float = 0.2, gamma: float = 1, kappa: float = 0.1, w_line: float = 0, w_edge: float = 0.4, max_iteration: int = 500):
    row, col = image.shape
    x, y = contour[:, 0], contour[:, 1]
    # 计算图像力的线函数与边函数
    Eline = image
    gx, gy = np.gradient(image)
    Eedge = -1*np.sqrt(gx*gx+gy*gy)

    # 外部力
    Eext = w_line*Eline + w_edge*Eedge

    # 计算五对角矩阵
    b1 = beta
    b2 = -(alpha+4*beta)
    b3 = 2*alpha+6*beta
    b4 = b2
    b5 = b1

    m = x.shape[0]
    A = b1*np.roll(np.eye(m), 2, axis=0)
    A = A + b2*np.roll(np.eye(m), 1, axis=0)
    A = A + b3*np.roll(np.eye(m), 0, axis=0)
    A = A + b4*np.roll(np.eye(m), -1, axis=0)
    A = A + b5*np.roll(np.eye(m), -2, axis=0)
    Ainv = np.linalg.inv(A+gamma*np.eye(m))

    # 迭代
    fx, fy = np.gradient(Eext)
    for i in range(max_iteration):
        ssx = gamma*x-kappa*(interp(fx, x, y))
        ssy = gamma*y-kappa*(interp(fy, x, y))
        x = Ainv @ ssx
        y = Ainv @ ssy
    return x, y

File: snake.m
clear,clc;
I = imread('IM000001_0005.png');
I = rgb2gray(I);
I = im2double(I);
Igs = imgaussfilt(I,1);
figure(1),imshow(Igs);  
x=[];y=[];c=1;N=100; %定义取点个数c,上限N  
% 获取User手动取点的坐标  
% [x,y]=getpts  
while c<N  
    [xi,yi,button]=ginput(1);  
    % 获取坐标向量  
    x=[x xi];  
    y=[y yi];  
    hold on  
    % text(xi,yi,'o','FontSize',10,'Color','red');  
    plot(xi,yi,'ro');  
    % 若为右击，则停止循环  
    if(button==3), break; end
    
    c=c+1;  
end

% 将第一个点复制到矩阵最后，构成Snake环  
xy = [x;y];  
c=c+1;  
xy(:,c)=xy(:,1);  
% 样条曲线差值  
t=1:c;  
ts = 1:0.1:c;  
xys = spline(t,xy,ts);  
xs = xys(1,:);  
ys = xys(2,:);  
% 样条差值效果  
hold on  
temp=plot(x(1),y(1),'ro',xs,ys,'b.');  
legend(temp,'原点','插值点');

% =========================================================================  
%                     Snakes算法实现部分  
% =========================================================================
NIter =2000; % 迭代次数  
alpha=0.2; beta=0.2; gamma = 1; kappa = 0.1;  
wl = 0; we=0.4; wt=0;  
[row ,col] = size(Igs);  
  
% 图像力-线函数  
Eline = Igs;  
% 图像力-边函数  
[gx,gy]=gradient(Igs);  
Eedge = -1*sqrt((gx.*gx+gy.*gy));  
% 图像力-终点函数  
% 卷积是为了求解偏导数，而离散点的偏导即差分求解  
m1 = [-1 1];   
m2 = [-1;1];  
m3 = [1 -2 1];   
m4 = [1;-2;1];  
m5 = [1 -1;-1 1];  
cx = conv2(Igs,m1,'same');  
cy = conv2(Igs,m2,'same');  
cxx = conv2(Igs,m3,'same');  
cyy = conv2(Igs,m4,'same');  
cxy = conv2(Igs,m5,'same');  
  
for i = 1:row  
    for j= 1:col  
        Eterm(i,j) = (cyy(i,j)*cx(i,j)*cx(i,j) -2 *cxy(i,j)*cx(i,j)*cy(i,j) + cxx(i,j)*cy(i,j)*cy(i,j))/((1+cx(i,j)*cx(i,j) + cy(i,j)*cy(i,j))^1.5);  
    end  
end  
  
%figure(3),imshow(Eterm);  
%figure(4),imshow(abs(Eedge));  
% 外部力 Eext = Eimage + Econ  
Eext = wl*Eline + we*Eedge + wt*Eterm;  
% 计算梯度  
[fx,fy]=gradient(Eext);  
  
xs=xs';  
ys=ys';  
[m, n] = size(xs);  
[mm, nn] = size(fx);  
  
% 计算五对角状矩阵  
% 附录: 公式（14） b(i)表示vi系数(i=i-2 到 i+2)  
b(1)=beta;  
b(2)=-(alpha + 4*beta);  
b(3)=(2*alpha + 6 *beta);  
b(4)=b(2);  
b(5)=b(1);  
  
A=b(1)*circshift(eye(m),2);  
A=A+b(2)*circshift(eye(m),1);  
A=A+b(3)*circshift(eye(m),0);  
A=A+b(4)*circshift(eye(m),-1);  
A=A+b(5)*circshift(eye(m),-2);  
  
% 计算矩阵的逆  
[L, U] = lu(A + gamma.* eye(m));  
Ainv = inv(U) * inv(L);

figure(3)
for i=1:NIter
ssx = gamma*xs - kappa*interp2(fx,xs,ys);
ssy = gamma*ys - kappa*interp2(fy,xs,ys);
xs = Ainv * ssx;
ys = Ainv*ssy;
imshow(I);
hold on;
plot([xs;xs(1)],[ys;ys(1)],'r-');
hold off;
pause(0.001)
end